In [2]:
# import libraries

import numpy as np
import pandas as pd

In [6]:
# read csv with pandas

df = pd.read_csv("D:/Medical Insurance Price Prediction/data/raw/insurance.csv")

df.head() # List the first 5 datapoints in the dataset

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), object(3)
memory usage: 73.3+ KB


- ### Has 7 Columns, 1338 data points, 6 independent features and 1 target feature (charges)

- ### Has 3 categorical columns

In [8]:
df.describe()

age          bmi     children       charges
count  1338.000000  1338.000000  1338.000000   1338.000000
mean     39.207025    30.663397     1.094918  13270.422265
std      14.049960     6.098187     1.205493  12110.011237
min      18.000000    15.960000     0.000000   1121.873900
25%      27.000000    26.296250     0.000000   4740.287150
50%      39.000000    30.400000     1.000000   9382.033000
75%      51.000000    34.693750     2.000000  16639.912515
max      64.000000    53.130000     5.000000  63770.428010

- ##### All the columns have non-null values and each row has a unique data point
- ##### Mean age is 39 years and each have at least 1 child and each person is charged a maximum of 13270

In [10]:
df.isnull().sum()

age         0
sex         0
bmi         0
children    0
smoker      0
region      0
charges     0
dtype: int64

- #### No nul values - not much cleaning needs to be done

In [14]:
# copy dataframe

new_df = df.copy()

In [ ]:
# one-hot encoding of categorical variables
# Sex
# Smoker
# region

# import OneHotEncoder from sklearn
from sklearn.preprocessing import OneHotEncoder

categorical_columns = ["sex", "smoker", "region"] # create a vraible and store the categorical variables in there

encoder = OneHotEncoder(sparse_output=False)

one_hot_encoded = encoder.fit_transform(new_df[categorical_columns])

one_hot_df = pd.DataFrame(one_hot_encoded, columns= encoder.get_feature_names_out(categorical_columns))

one_hot_df

sex_female  sex_male  smoker_no  smoker_yes  region_northeast  \
0            1.0       0.0        0.0         1.0               0.0   
1            0.0       1.0        1.0         0.0               0.0   
2            0.0       1.0        1.0         0.0               0.0   
3            0.0       1.0        1.0         0.0               0.0   
4            0.0       1.0        1.0         0.0               0.0   
...          ...       ...        ...         ...               ...   
1333         0.0       1.0        1.0         0.0               0.0   
1334         1.0       0.0        1.0         0.0               1.0   
1335         1.0       0.0        1.0         0.0               0.0   
1336         1.0       0.0        1.0         0.0               0.0   
1337         1.0       0.0        0.0         1.0               0.0   

      region_northwest  region_southeast  region_southwest  
0                  0.0               0.0               1.0  
1                  0.0               1.0               0.0  
2                  0.0               1.0               0.0  
3                  1.0               0.0               0.0  
4                  1.0               0.0               0.0  
...                ...               ...               ...  
1333               1.0               0.0               0.0  
1334               0.0               0.0               0.0  
1335               0.0               1.0               0.0  
1336               0.0               0.0               1.0  
1337               1.0               0.0               0.0  

[1338 rows x 8 columns]

In [19]:
# Concatenate with the rest of the columns

df_encoded = pd.concat([new_df, one_hot_df], axis=1)

# Drop the orignal categorical columns
df_encoded = df_encoded.drop(categorical_columns, axis=1)

df_encoded

age     bmi  children      charges  sex_female  sex_male  smoker_no  \
0      19  27.900         0  16884.92400         1.0       0.0        0.0   
1      18  33.770         1   1725.55230         0.0       1.0        1.0   
2      28  33.000         3   4449.46200         0.0       1.0        1.0   
3      33  22.705         0  21984.47061         0.0       1.0        1.0   
4      32  28.880         0   3866.85520         0.0       1.0        1.0   
...   ...     ...       ...          ...         ...       ...        ...   
1333   50  30.970         3  10600.54830         0.0       1.0        1.0   
1334   18  31.920         0   2205.98080         1.0       0.0        1.0   
1335   18  36.850         0   1629.83350         1.0       0.0        1.0   
1336   21  25.800         0   2007.94500         1.0       0.0        1.0   
1337   61  29.070         0  29141.36030         1.0       0.0        0.0   

      smoker_yes  region_northeast  region_northwest  region_southeast  \
0            1.0               0.0               0.0               0.0   
1            0.0               0.0               0.0               1.0   
2            0.0               0.0               0.0               1.0   
3            0.0               0.0               1.0               0.0   
4            0.0               0.0               1.0               0.0   
...          ...               ...               ...               ...   
1333         0.0               0.0               1.0               0.0   
1334         0.0               1.0               0.0               0.0   
1335         0.0               0.0               0.0               1.0   
1336         0.0               0.0               0.0               0.0   
1337         1.0               0.0               1.0               0.0   

      region_southwest  
0                  1.0  
1                  0.0  
2                  0.0  
3                  0.0  
4                  0.0  
...                ...  
1333               0.0  
1334               0.0  
1335               0.0  
1336               1.0  
1337               0.0  

[1338 rows x 12 columns]

In [20]:
df_encoded.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   age               1338 non-null   int64  
 1   bmi               1338 non-null   float64
 2   children          1338 non-null   int64  
 3   charges           1338 non-null   float64
 4   sex_female        1338 non-null   float64
 5   sex_male          1338 non-null   float64
 6   smoker_no         1338 non-null   float64
 7   smoker_yes        1338 non-null   float64
 8   region_northeast  1338 non-null   float64
 9   region_northwest  1338 non-null   float64
 10  region_southeast  1338 non-null   float64
 11  region_southwest  1338 non-null   float64
dtypes: float64(10), int64(2)
memory usage: 125.6 KB


In [21]:
# Save df_encoded as csv
df_encoded.to_csv("one_hot_encoded_df")